$\large{\textbf{2.1 Formulation:}} \\ \text{Let } c_{i,j} \text{ be the cost of setting } i^{th} \text{ factory at the } j^{th} \text{ location.} \\ \text{Let } x_{i,j} \text{ be the decision variables which takes value 1 if ith factory is set-up at jth location, else 0.} \\ \text{Let No. of factories = No. of locations = n} \\ min \ \sum_{i=1}^{n} ∑_{j=1}^{n} c_{i,j} \ x_{i,j} \\ s.t. \ \sum_{j=1}^{n} x_{i,j} =1 \ \  ∀ \ i=\{1,2,...,n\} \ \textbf{(Each factory can be set-up at only 1 location)}  \\ \  \sum_{i=1}^{n} x_{i,j} =1 \ \  ∀ \ j=\{1,2,...,n\} \ \textbf{(Each location can have only 1 factory set-up)} \\ x_{i,j} \in \{0,1\} \ \ ∀ \ i,j$

In [1]:
!pip install -q pyomo

     |████████████████████████████████| 9.7 MB 3.7 MB/s 
     |████████████████████████████████| 49 kB 5.5 MB/s 


In [2]:
from pyomo.environ import *

In [3]:
import numpy as np

In [4]:
data = np.loadtxt('lab6_ex2.txt', delimiter = '   ')

In [5]:
data

array([[ 0.,  1.,  2.,  3.,  4.,  5.,  6.,  7.,  8.,  9., 10., 11., 12.],
       [ 1., 21., 25., 18., 19., 23., 24., 16., 17., 21., 23., 19., 21.],
       [ 2., 19., 22., 19., 19., 23., 21., 15., 20., 16., 17., 17., 24.],
       [ 3., 21., 19., 20., 18., 22., 24., 14., 16., 18., 16., 20., 24.],
       [ 4., 19., 22., 17., 17., 22., 22., 18., 18., 20., 21., 23., 23.],
       [ 5., 18., 22., 15., 15., 23., 23., 21., 25., 24., 19., 21., 23.],
       [ 6., 23., 19., 19., 14., 24., 21., 23., 21., 20., 20., 22., 19.],
       [ 7., 21., 16., 27., 17., 26., 27., 16., 19., 24., 20., 20., 23.],
       [ 8., 22., 17., 21., 15., 23., 24., 24., 20., 22., 19., 23., 21.],
       [ 9., 23., 24., 18., 19., 22., 23., 21., 18., 23., 15., 25., 21.],
       [10., 18., 21., 21., 18., 26., 24., 18., 17., 18., 20., 21., 20.],
       [11., 19., 12., 21., 19., 23., 20., 22., 19., 17., 19., 20., 24.],
       [12., 18., 22., 24., 17., 21., 28., 18., 16., 22., 24., 25., 24.]])

In [6]:
model_ex2 = ConcreteModel()

In [7]:
data.shape

(13, 13)

In [8]:
#No of factories
M = data.shape[0] - 1

#No of locations
N = data.shape[1] - 1 

In [9]:
row_indices = np.arange(M)
col_indices = np.arange(N)

In [10]:
model_ex2.x = Var(row_indices, col_indices, domain = Binary)

In [11]:
model_ex2.obj = Objective(expr = sum(data[i+1,j+1]*model_ex2.x[i,j] for i in row_indices for j in col_indices), sense = minimize)

In [12]:
model_ex2.const = ConstraintList()

In [13]:
#Factory Constraint
for i in row_indices:
  model_ex2.const.add(expr = sum(model_ex2.x[i,j] for j in col_indices) == 1)


#Location Constraits:
for j in col_indices:
  model_ex2.const.add(expr = sum(model_ex2.x[i,j] for i in row_indices) == 1)

In [14]:
model_ex2.pprint()

4 Set Declarations
    const_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   24 : {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain              : Size : Members
        None :     2 : x_index_0*x_index_1 :  144 : {(0, 0), (0, 1), (0, 2), (0, 3), (0, 4), (0, 5), (0, 6), (0, 7), (0, 8), (0, 9), (0, 10), (0, 11), (1, 0), (1, 1), (1, 2), (1, 3), (1, 4), (1, 5), (1, 6), (1, 7), (1, 8), (1, 9), (1, 10), (1, 11), (2, 0), (2, 1), (2, 2), (2, 3), (2, 4), (2, 5), (2, 6), (2, 7), (2, 8), (2, 9), (2, 10), (2, 11), (3, 0), (3, 1), (3, 2), (3, 3), (3, 4), (3, 5), (3, 6), (3, 7), (3, 8), (3, 9), (3, 10), (3, 11), (4, 0), (4, 1), (4, 2), (4, 3), (4, 4), (4, 5), (4, 6), (4, 7), (4, 8), (4, 9), (4, 10), (4, 11), (5, 0), (5, 1), (5, 2), (5, 3), (5, 4), (5, 5), (5, 6), (5, 7), (5, 8), (5, 9), (5, 10), (5, 11), (6

In [15]:
!apt-get install -y -qq coinor-cbc

Selecting previously unselected package coinor-libcoinutils3v5.
(Reading database ... 155569 files and directories currently installed.)
Preparing to unpack .../0-coinor-libcoinutils3v5_2.10.14+repack1-1_amd64.deb ...
Unpacking coinor-libcoinutils3v5 (2.10.14+repack1-1) ...
Selecting previously unselected package coinor-libosi1v5.
Preparing to unpack .../1-coinor-libosi1v5_0.107.9+repack1-1_amd64.deb ...
Unpacking coinor-libosi1v5 (0.107.9+repack1-1) ...
Selecting previously unselected package coinor-libclp1.
Preparing to unpack .../2-coinor-libclp1_1.16.11+repack1-1_amd64.deb ...
Unpacking coinor-libclp1 (1.16.11+repack1-1) ...
Selecting previously unselected package coinor-libcgl1.
Preparing to unpack .../3-coinor-libcgl1_0.59.10+repack1-1_amd64.deb ...
Unpacking coinor-libcgl1 (0.59.10+repack1-1) ...
Selecting previously unselected package coinor-libcbc3.
Preparing to unpack .../4-coinor-libcbc3_2.9.9+repack1-1_amd64.deb ...
Unpacking coinor-libcbc3 (2.9.9+repack1-1) ...
Selecting p

In [16]:
opt_cbc = SolverFactory('cbc')

In [17]:
result = opt_cbc.solve(model_ex2)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [18]:
# display solution
print('\nObjective = ', model_ex2.obj())

print('\nDecision Variables')
for i in row_indices:
  for j in col_indices:
    if(model_ex2.x[i,j].value != 0):
      print('{}th factory is located on the {}th location'.format(i+1,j+1))

print('\nConstraints')
model_ex2.const.display()


Objective =  204.0

Decision Variables
1th factory is located on the 11th location
2th factory is located on the 9th location
3th factory is located on the 8th location
4th factory is located on the 6th location
5th factory is located on the 3th location
6th factory is located on the 12th location
7th factory is located on the 7th location
8th factory is located on the 4th location
9th factory is located on the 10th location
10th factory is located on the 1th location
11th factory is located on the 2th location
12th factory is located on the 5th location

Constraints
const : Size=24
    Key : Lower : Body : Upper
      1 :   1.0 :  1.0 :   1.0
      2 :   1.0 :  1.0 :   1.0
      3 :   1.0 :  1.0 :   1.0
      4 :   1.0 :  1.0 :   1.0
      5 :   1.0 :  1.0 :   1.0
      6 :   1.0 :  1.0 :   1.0
      7 :   1.0 :  1.0 :   1.0
      8 :   1.0 :  1.0 :   1.0
      9 :   1.0 :  1.0 :   1.0
     10 :   1.0 :  1.0 :   1.0
     11 :   1.0 :  1.0 :   1.0
     12 :   1.0 :  1.0 :   1.0
     1

$\large{\textbf{Answer 2.8:}} \\ \textbf{Total cost = Rs. 204} \\ \text{Factory} → \text{Location} \\ 1 → 11 \\2 → 9 \\3 → 8 \\4 → 6 \\ 5 → 3 \\6 → 12 \\7 → 7 \\8 → 4 \\9 → 10 \\10 → 1 \\11 → 2 \\12 → 5 $

$\large{\textbf{Answer 2.9}}$

In [19]:
model_ex2.x.domain=NonNegativeReals

In [20]:
result = opt_cbc.solve(model_ex2)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [21]:
# display solution
print('\nObjective = ', model_ex2.obj())

print('\nDecision Variables')
for i in row_indices:
  for j in col_indices:
    if(model_ex2.x[i,j].value != 0):
      print(model_ex2.x[i,j].value)
      print('{}th factory is located on the {}th location'.format(i+1,j+1))


print('\nConstraints')
model_ex2.const.display()


Objective =  204.0

Decision Variables
1.0
1th factory is located on the 11th location
1.0
2th factory is located on the 9th location
1.0
3th factory is located on the 8th location
1.0
4th factory is located on the 6th location
1.0
5th factory is located on the 3th location
1.0
6th factory is located on the 12th location
1.0
7th factory is located on the 7th location
1.0
8th factory is located on the 4th location
1.0
9th factory is located on the 10th location
1.0
10th factory is located on the 1th location
1.0
11th factory is located on the 2th location
1.0
12th factory is located on the 5th location

Constraints
const : Size=24
    Key : Lower : Body : Upper
      1 :   1.0 :  1.0 :   1.0
      2 :   1.0 :  1.0 :   1.0
      3 :   1.0 :  1.0 :   1.0
      4 :   1.0 :  1.0 :   1.0
      5 :   1.0 :  1.0 :   1.0
      6 :   1.0 :  1.0 :   1.0
      7 :   1.0 :  1.0 :   1.0
      8 :   1.0 :  1.0 :   1.0
      9 :   1.0 :  1.0 :   1.0
     10 :   1.0 :  1.0 :   1.0
     11 :   1.0 :  1

$\textbf{Total cost = Rs. 204} \\ \text{Factory} → \text{Location} \\ 1 → 11 \\2 → 9 \\3 → 8 \\4 → 6 \\ 5 → 3 \\6 → 12 \\7 → 7 \\8 → 4 \\9 → 10 \\10 → 1 \\11 → 2 \\12 → 5 $

$\large{\textbf{Answer 2.10: }} \\ \text{Yes, the optimal cost in both the cases comes out to be same and the values of the decision variables are integers as well.} \\  \text{This is because here we are taking a convex combination of costs along the rows as well as the columns. (because the sum of the coefficients of the costs sums to 1)}\\ \text{And we know convex combination of n numbers is greater than or equal to the minimum of those n values.} \\ \text{So total cost is minimized when the value of the convex combination of the row/column is equal to the minimum cost along the row/column.} \\ \text{And the minimum value is attained when the coefficient of the least value is 1 and all other coefficients are 0.}$

In [22]:
data = data + 0.1

In [23]:
data

array([[ 0.1,  1.1,  2.1,  3.1,  4.1,  5.1,  6.1,  7.1,  8.1,  9.1, 10.1,
        11.1, 12.1],
       [ 1.1, 21.1, 25.1, 18.1, 19.1, 23.1, 24.1, 16.1, 17.1, 21.1, 23.1,
        19.1, 21.1],
       [ 2.1, 19.1, 22.1, 19.1, 19.1, 23.1, 21.1, 15.1, 20.1, 16.1, 17.1,
        17.1, 24.1],
       [ 3.1, 21.1, 19.1, 20.1, 18.1, 22.1, 24.1, 14.1, 16.1, 18.1, 16.1,
        20.1, 24.1],
       [ 4.1, 19.1, 22.1, 17.1, 17.1, 22.1, 22.1, 18.1, 18.1, 20.1, 21.1,
        23.1, 23.1],
       [ 5.1, 18.1, 22.1, 15.1, 15.1, 23.1, 23.1, 21.1, 25.1, 24.1, 19.1,
        21.1, 23.1],
       [ 6.1, 23.1, 19.1, 19.1, 14.1, 24.1, 21.1, 23.1, 21.1, 20.1, 20.1,
        22.1, 19.1],
       [ 7.1, 21.1, 16.1, 27.1, 17.1, 26.1, 27.1, 16.1, 19.1, 24.1, 20.1,
        20.1, 23.1],
       [ 8.1, 22.1, 17.1, 21.1, 15.1, 23.1, 24.1, 24.1, 20.1, 22.1, 19.1,
        23.1, 21.1],
       [ 9.1, 23.1, 24.1, 18.1, 19.1, 22.1, 23.1, 21.1, 18.1, 23.1, 15.1,
        25.1, 21.1],
       [10.1, 18.1, 21.1, 21.1, 18.1, 26.1, 24.1, 

In [24]:
model_ex2.new_obj = Objective(expr = sum(data[i+1,j+1]*model_ex2.x[i,j] for i in row_indices for j in col_indices), sense = minimize)

In [25]:
model_ex2.obj.deactivate()

In [26]:
result = opt_cbc.solve(model_ex2)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [27]:
# display solution
print('\nObjective = ', model_ex2.new_obj())

print('\nDecision Variables')
for i in row_indices:
  for j in col_indices:
    if(model_ex2.x[i,j].value != 0):
      print(model_ex2.x[i,j].value)
      print('{}th factory is located on the {}th location'.format(i+1,j+1))

print('\nConstraints')
model_ex2.const.display()


Objective =  205.19999999999996

Decision Variables
1.0
1th factory is located on the 11th location
1.0
2th factory is located on the 9th location
1.0
3th factory is located on the 8th location
1.0
4th factory is located on the 6th location
1.0
5th factory is located on the 3th location
1.0
6th factory is located on the 12th location
1.0
7th factory is located on the 7th location
1.0
8th factory is located on the 4th location
1.0
9th factory is located on the 10th location
1.0
10th factory is located on the 1th location
1.0
11th factory is located on the 2th location
1.0
12th factory is located on the 5th location

Constraints
const : Size=24
    Key : Lower : Body : Upper
      1 :   1.0 :  1.0 :   1.0
      2 :   1.0 :  1.0 :   1.0
      3 :   1.0 :  1.0 :   1.0
      4 :   1.0 :  1.0 :   1.0
      5 :   1.0 :  1.0 :   1.0
      6 :   1.0 :  1.0 :   1.0
      7 :   1.0 :  1.0 :   1.0
      8 :   1.0 :  1.0 :   1.0
      9 :   1.0 :  1.0 :   1.0
     10 :   1.0 :  1.0 :   1.0
     11

$\large{\textbf{Answer 2.11:}} \\ \text{Even after takings costs as reals, we get the same result. This implies that the data type of costs matrix makes no change in the optimal solution of the assignment problem.} \\ \text{Only the total costs change but the assignment remains the same.}$

In [28]:
model_ex2.obj.deactivate()

In [29]:
model_ex2.new_obj.deactivate()

In [30]:
data -= 0.1

In [31]:
data[3,4] = 500
data[9,6] = 500
data[4,10] = 500

In [32]:
model_ex2.new_obj2 = Objective(expr = sum(data[i+1,j+1]*model_ex2.x[i,j] for i in row_indices for j in col_indices), sense = minimize)

In [33]:
model_ex2.x.domain=Binary

In [34]:
model_ex2.pprint()

4 Set Declarations
    const_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   24 : {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain              : Size : Members
        None :     2 : x_index_0*x_index_1 :  144 : {(0, 0), (0, 1), (0, 2), (0, 3), (0, 4), (0, 5), (0, 6), (0, 7), (0, 8), (0, 9), (0, 10), (0, 11), (1, 0), (1, 1), (1, 2), (1, 3), (1, 4), (1, 5), (1, 6), (1, 7), (1, 8), (1, 9), (1, 10), (1, 11), (2, 0), (2, 1), (2, 2), (2, 3), (2, 4), (2, 5), (2, 6), (2, 7), (2, 8), (2, 9), (2, 10), (2, 11), (3, 0), (3, 1), (3, 2), (3, 3), (3, 4), (3, 5), (3, 6), (3, 7), (3, 8), (3, 9), (3, 10), (3, 11), (4, 0), (4, 1), (4, 2), (4, 3), (4, 4), (4, 5), (4, 6), (4, 7), (4, 8), (4, 9), (4, 10), (4, 11), (5, 0), (5, 1), (5, 2), (5, 3), (5, 4), (5, 5), (5, 6), (5, 7), (5, 8), (5, 9), (5, 10), (5, 11), (6

In [35]:
result = opt_cbc.solve(model_ex2)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [36]:
# display solution
print('\nObjective = ', model_ex2.new_obj2())

print('\nDecision Variables')
for i in row_indices:
  for j in col_indices:
    if(model_ex2.x[i,j].value != 0):
      print('{}th factory is located on the {}th location'.format(i+1,j+1))

print('\nConstraints')
model_ex2.const.display()


Objective =  204.0

Decision Variables
1th factory is located on the 11th location
2th factory is located on the 9th location
3th factory is located on the 8th location
4th factory is located on the 6th location
5th factory is located on the 3th location
6th factory is located on the 12th location
7th factory is located on the 7th location
8th factory is located on the 4th location
9th factory is located on the 10th location
10th factory is located on the 1th location
11th factory is located on the 2th location
12th factory is located on the 5th location

Constraints
const : Size=24
    Key : Lower : Body : Upper
      1 :   1.0 :  1.0 :   1.0
      2 :   1.0 :  1.0 :   1.0
      3 :   1.0 :  1.0 :   1.0
      4 :   1.0 :  1.0 :   1.0
      5 :   1.0 :  1.0 :   1.0
      6 :   1.0 :  1.0 :   1.0
      7 :   1.0 :  1.0 :   1.0
      8 :   1.0 :  1.0 :   1.0
      9 :   1.0 :  1.0 :   1.0
     10 :   1.0 :  1.0 :   1.0
     11 :   1.0 :  1.0 :   1.0
     12 :   1.0 :  1.0 :   1.0
     1

$\large{\textbf{Answer 2.12:}} \\ \textbf{Total cost = Rs. 204} \\ \text{Factory} → \text{Location} \\ 1 → 11 \\2 → 9 \\3 → 8 \\4 → 6 \\ 5 → 3 \\6 → 12 \\7 → 7 \\8 → 4 \\9 → 10 \\10 → 1 \\11 → 2 \\12 → 5 \\ \text{Here we can observe that the optimal cost remains the same as the original model because when the optimal assignment was done,} \\ \text{no illegal assignments(according to our new condition) was done, so the new condition has no impact on the optimal cost and the assignment.} $